synapse/tests/test_cortex.py

```diff
@@ -7578,3 +7578,217 @@ async def test_cortex_dmons_after_modelrev(self):
+            self.isin(vtype, core.vaulttypes)
+
+            # visi1 can create visi1 vaults
+            uiden = await core.addVault('user', 'visi1', vtype, {'apikey': 'test1234'}, user=visi1)
+            self.isin(f'user:visi1:{vtype}', core.vaultsbyname)
+            self.isin(uiden, core.vaultsbyiden)
+
+            vault = await core.getVault(uiden)
+            self.nn(vault)
+            self.eq(vault['ident'], visi1.iden)
+            self.eq(vault['data'], {'apikey': 'test1234'})
+
+            # visi1 is a contributor so he can create contributor vaults
+            riden = await core.addVault('role', 'contributor', vtype, {'apikey': 'test2345'}, user=visi1)
+            self.isin(f'role:contributor:{vtype}', core.vaultsbyname)
+            self.isin(riden, core.vaultsbyiden)
+
+            vault = await core.getVault(riden)
+            self.nn(vault)
+            self.eq(vault['ident'], contributor.iden)
+            self.eq(vault['data'], {'apikey': 'test2345'})
+
+            # Admin can create global vaults
+            giden = await core.addVault('global', None, vtype, {'apikey': 'test3456'}, user=root)
+            self.isin(f'global::{vtype}', core.vaultsbyname)
+            self.isin(giden, core.vaultsbyiden)
+
+            vault = await core.getVault(giden)
+            self.nn(vault)
+            self.none(vault['ident'])
+            self.eq(vault['data'], {'apikey': 'test3456'})
+
+            # Change the data in each of the vaults
+            await core.setVault(uiden, {'apikey': 'foo1234'})
+            vault = await core.getVault(uiden)
+            self.eq(vault['data'], {'apikey': 'foo1234'})
+
+            await core.setVault(riden, {'apikey': 'foo2345'})
+            vault = await core.getVault(riden)
+            self.eq(vault['data'], {'apikey': 'foo2345'})
+
+            await core.setVault(giden, {'apikey': 'foo3456'})
+            vault = await core.getVault(giden)
+            self.eq(vault['data'], {'apikey': 'foo3456'})
+
+            # Set default vault to global and open vault
+            await core.setVaultTypeDefault(vtype, 'global')
+            iden = await core.openVault(vtype, user=visi1)
+            self.eq(iden, giden)
+
+            # Set default vault to role and open vault
+            await core.setVaultTypeDefault(vtype, 'role')
+            iden = await core.openVault(vtype, user=visi1)
+            self.eq(iden, riden)
+
+            iden = await core.openVault(vtype, scope='user', user=visi1)
+            self.eq(iden, uiden)
+
+            # Root sees all
+            vaults = [k async for k in core.listVaults()]
+            self.len(3, vaults)
+            self.isin((uiden, f'user:visi1:{vtype}', vtype, 'user'), vaults)
+            self.isin((riden, f'role:contributor:{vtype}', vtype, 'role'), vaults)
+            self.isin((giden, f'global::{vtype}', vtype, 'global'), vaults)
+
+            # visi2 can't list any vaults because he only has PERM_READ to global
```

Comment:
This comment is wrong. visi2 can see the global vault only because he has PERM_READ to it.